### PR TITLE
Constant Fold Logf128 calls

### PR DESCRIPTION
**Pull request by MDevereau**
This patch enables constant folding for 128 bit floating-point logf calls. This is achieved by querying if the host system has the logf128() symbol available with a CMake test. If so, replace the runtime call with the compile time value returned from logf128.

### CODE REVIEW

**Comment by MDevereau** [comment]
To give some context as to why this is desirable, there's been instances where calls to long double have been generated when they should be constant folded. As they are 128-bit floats on aarch64 these calls end up being particularly slow compared to targets where the `long double` type is 64 bits in length.

The mersenne twister in libstdc++ contains this code:
```c++
const long double __r = static_cast<long double>(__urng.max())        
                      - static_cast<long double>(__urng.min()) + 1.0L;
const size_t __log2r = std::log(__r) / std::log(2.0L);  
```
GCC will constant fold both calls to logl here but clang will keep the calls. The proposed approach in the pull request is in thoery an easy-win which other systems could also benefit from.

**Comment by efriedma-quic** [comment]
The primary reason we haven't continued to expand float constant-folding is that we don't really want to use the host's libm for target calculations.  Ideally, we want the compiler to be portable: we want to consistently produce the same results.  So we want to use our own library routines.

We have ConstantFoldFP which already uses the host libc, but we don't really want to expand that usage.  gcc uses MFPR, but LLVM can't due to licensing issues.  Maybe we can borrow code from llvm-libc?

**Comment by efriedma-quic** [comment]
> This approach relies on unit tests, as more commonplace Clang/C tests and opt/llc/IR tests are not applicable since they are ignorant to the result of the compile time CMake check.

You can write code in lit.cfg.py to add a feature to config.available_features, then check for it using a REQUIRES clause.

File: llvm/lib/Analysis/ConstantFolding.cpp [hunk 1]
```diff
@@ -2209,6 +2209,11 @@ static Constant *ConstantFoldScalarCall1(StringRef Name,
     switch (IntrinsicID) {
       default: break;
       case Intrinsic::log:
+#if defined(__FLOAT128__) && defined(HAS_LOGF128)
+        if (Ty->isFP128Ty()) {
+          return ConstantFP::get(Ty, logf128(APF.convertToQuad()));
+        }
+#endif
         return ConstantFoldFP(log, APF, Ty);
```
Comment (efriedma-quic) [inline llvm/lib/Analysis/ConstantFolding.cpp:2217]: This falls through to using the double-precision log() for long double values on hosts which don't have float128; that seems wrong.

File: llvm/include/llvm/Support/float128.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_FLOAT128
+#define LLVM_FLOAT128
+
+#if defined(__clang__) && defined(__FLOAT128__)
+typedef __float128 float128;
```
Comment (efriedma-quic) [inline llvm/include/llvm/Support/float128.h:13]: `namespace llvm`?

File: llvm/unittests/Analysis/ConstantLogf128.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
+//===- unittests/CodeGen/BufferSourceTest.cpp - MemoryBuffer source tests -===//
```
Comment (arsenm) [inline llvm/unittests/Analysis/ConstantLogf128.cpp:1]: Copy paste error

File: llvm/unittests/Analysis/ConstantLogf128.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+#include "llvm/CodeGen/GlobalISel/CallLowering.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstrTypes.h"
+#include "gtest/gtest.h"
+
+using namespace llvm;
+
+namespace {
+
+TEST(ConstantFoldLogf128Fixture, ConstantFoldLogf128) {
```
Comment (arsenm) [inline llvm/unittests/Analysis/ConstantLogf128.cpp:20]: Easier to write InstSimplify tests for this. Also should test all the edge case special values

File: llvm/include/llvm/IR/Constants.h [hunk 5]
```diff
@@ -289,6 +289,10 @@ class ConstantFP final : public ConstantData {
   /// host double and as the target format.
   static Constant *get(Type *Ty, double V);
 
+#ifdef __FLOAT128__
+  static Constant *get128(Type *Ty, float128 V);
```
Comment (arsenm) [inline llvm/include/llvm/IR/Constants.h:293]: Just use the APFloat version for float128?

File: llvm/lib/Analysis/ConstantFolding.cpp (hunk 1 above)
Comment (MDevereau) [reply llvm/lib/Analysis/ConstantFolding.cpp:2217]: Is this not already what was happening previously? Inside `ConstantFoldFP` the result is handled as a double. I'm not sure if you're asking me to start preserving the long double type in this pre-existing code or whether what I've introduced has changed established behaviour

Comment (efriedma-quic) [reply llvm/lib/Analysis/ConstantFolding.cpp:2217]: Without this patch, we wouldn't try to fold fp128 at all.  With this patch, we try to fold it using "double" routines.  (Previously, we would only fold 64-bit or 32-bit floats.)

File: llvm/include/llvm/ADT/APInt.h [hunk 6]
```diff
@@ -1688,6 +1696,16 @@ class [[nodiscard]] APInt {
     return APInt(sizeof(float) * CHAR_BIT, llvm::bit_cast<uint32_t>(V));
   }
 
+#ifdef __FLOAT128__
+  static APInt longDoubleToBits(float128 V) {
```
Comment (arsenm) [inline llvm/include/llvm/ADT/APInt.h:1700]: I don't think APInt/APFloat should be exposing APIs that are only conditionally available. That kind of defeats the point of using them

**Comment by jcranmer-intel** [comment]
> The primary reason we haven't continued to expand float constant-folding is that we don't really want to use the host's libm for target calculations. Ideally, we want the compiler to be portable: we want to consistently produce the same results. So we want to use our own library routines.
> 
> We have ConstantFoldFP which already uses the host libc, but we don't really want to expand that usage. gcc uses MFPR, but LLVM can't due to licensing issues. Maybe we can borrow code from llvm-libc?

I whole-heartedly agree with this sentiment. llvm-libc is working on getting correctly-rounded versions of all the math functions, but the only one implemented for all the float sizes is `sqrt`, and otherwise they're largely implemented only for `float` at the moment.

When we talked at the last LLVM FP working group meeting, llvm-libc people were receptive to the idea of reusing code for APFloat, although I would probably want to wait to see how things work out for reusing libc code in libc++ first.

File: llvm/include/llvm/ADT/APInt.h (hunk 6 above)
Comment (MDevereau) [reply llvm/include/llvm/ADT/APInt.h:1700]: I agree, but I can't think of a better way to do this currently since logf128 isn't available on all targets but is the only type that guarantees a floating-point type of 128 bits.

File: llvm/include/llvm/Support/float128.h (hunk 2 above)
Comment (MDevereau) [reply llvm/include/llvm/Support/float128.h:15]: Added.

File: llvm/include/llvm/IR/Constants.h (hunk 5 above)
Comment (MDevereau) [reply llvm/include/llvm/IR/Constants.h]: Sure, I've removed this function now.

File: llvm/unittests/Analysis/ConstantLogf128.cpp (hunk 4 above)
Comment (MDevereau) [reply llvm/unittests/Analysis/ConstantLogf128.cpp]: I've removed this file and moved/added new tests to InstSimplify. Hopefully that should prove easier to test.

File: llvm/unittests/Analysis/ConstantLogf128.cpp (hunk 3 above)
Comment (MDevereau) [reply llvm/unittests/Analysis/ConstantLogf128.cpp]: Oops. The file has been removed now though.

File: llvm/lib/Analysis/ConstantFolding.cpp (hunk 1 above)
Comment (MDevereau) [reply llvm/lib/Analysis/ConstantFolding.cpp]: Ah I see what you mean now, thanks for clarifying. I've rearranged this to stop any fall-through happening

**Comment by MDevereau** [comment]
> You can write code in lit.cfg.py to add a feature to config.available_features, then check for it using a REQUIRES clause.

I've implemented this and removed the unit tests. Instead I'm now relying on InstSimplify tests.

File: llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  ret fp128 %A
+}
+
+define  fp128 @log_e_minus_2(){
+; CHECK-LABEL: define fp128 @log_e_minus_2() {
+; CHECK-NEXT:    ret fp128 0xL00000000000000007FFF800000000000
+;
+  %A = call fp128 @llvm.log.f128(fp128 noundef 0xL0000000000000000C000000000000000)
+  ret fp128 %A
+}
```
Comment (arsenm) [inline llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll:79]: test inf/nan/zero

Comment (MDevereau) [reply llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll:79]: Added.

File: llvm/include/llvm/ADT/APInt.h (hunk 6 above)
Comment (efriedma-quic) [reply llvm/include/llvm/ADT/APInt.h:1700]: The libc logf128() function takes a float128 argument, and returns a float128 result.  If we're using it, we need to convert between float128 and APFloat.  (If APFloat::log() existed, this patch would be much simpler... but it doesn't, and it's difficult to write.)

File: llvm/include/llvm/ADT/APFloat.h [hunk 8]
```diff
@@ -299,6 +300,9 @@ class IEEEFloat final : public APFloatBase {
   IEEEFloat(const fltSemantics &, integerPart);
   IEEEFloat(const fltSemantics &, uninitializedTag);
   IEEEFloat(const fltSemantics &, const APInt &);
+#ifdef __FLOAT128__
+  explicit IEEEFloat(float128 ld);
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/include/llvm/ADT/APFloat.h]: Removed.

File: llvm/include/llvm/ADT/APInt.h (hunk 6 above)
Comment (MDevereau) [reply llvm/include/llvm/ADT/APInt.h]: I think it's correct to remove this (and have done so) since APFloat already has internal representation of float128's. I'm not sure about the `convertToQuad` function I've added though, since that is necessary to extract the float128 information from APFloat in order to pass it to the C logf128 function.

File: llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll [hunk 9]
```diff
@@ -70,10 +70,50 @@ define  fp128 @log_e_smallest_number_larger_than_one(){
...
+
+define  fp128 @log_e_0(){
+; CHECK-LABEL: define fp128 @log_e_0() {
+; CHECK-NEXT:    ret fp128 0xL0000000000000000FFFF000000000000
+;
+  %A = call fp128 @llvm.log.f128(fp128 noundef 0xL00000000000000000000000000000000)
+  ret fp128 %A
+}
+
+define  fp128 @log_e_negative_0(){
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll]: Removed

File: llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll [hunk 10]
```diff
@@ -70,10 +70,50 @@ define  fp128 @log_e_smallest_number_larger_than_one(){
...
+  ret fp128 %A
+}
+
+define  fp128 @log_e_nan(){
+; CHECK-LABEL: define fp128 @log_e_nan() {
+; CHECK-NEXT:    ret fp128 0xL00000000000000007FFF800000000001
+;
+  %A = call fp128 @llvm.log.f128(fp128 noundef 0xL00000000000000007FFF000000000001)
+  ret fp128 %A
+}
```
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll:119]: I'm not sure what you mean, sorry. I initially thought you meant the log function can take a vector type parameter, but after checking https://llvm.org/docs/LangRef.html#llvm-log-intrinsic I can't see one.

File: llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll [hunk 11]
```diff
@@ -0,0 +1,119 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt < %s -passes=instsimplify -S | FileCheck %s
+
+; REQUIRES: has_logf128
+; REQUIRES: aarch64-registered-target
```
Comment (arsenm) [inline llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll:5]: This doesn't require aarch64

**Review by arsenm — COMMENTED** [review]
LGTM except for the unnecessary target requires

File: llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll (hunk 10 above)
[the remark this answers is not in the document]
Comment (MDevereau) [reply llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll:118]: Thanks, I've added that test now.

File: llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll (hunk 11 above)
Comment (MDevereau) [reply llvm/test/Transforms/InstSimplify/ConstProp/logf128.ll]: Removed. The initial reasoning behind this was that I was unsure if it was safe to use these tests for long double on multiple targets, like on x86 where the width may be 80 bits instead of the full 128 bits like on AArch64.

**Comment by efriedma-quic** [comment]
Please fix commit message; I think you've addressed most of the "concerns".

I'm still not happy with the dependency on host libc, given the increasing pressure to have constant-folding functionality reliably available.  See also https://discourse.llvm.org/t/rfc-project-hand-in-hand-llvm-libc-libc-code-sharing/77701/18 .

Other than that, looks fine.

**Comment by MDevereau** [comment]
@efriedma-quic I've changed the commit message to be brief and removed any WIP-like language

File: llvm/lib/Analysis/CMakeLists.txt [hunk 12]
```diff
@@ -161,3 +161,9 @@ add_llvm_component_library(LLVMAnalysis
   Support
   TargetParser
   )
+
+include(CheckCXXSymbolExists)
+check_cxx_symbol_exists(logf128 math.h HAS_LOGF128)
+if(HAS_LOGF128)
+ target_compile_definitions(LLVMAnalysis PRIVATE HAS_LOGF128)
+endif()
```
Comment (arsenm) [inline llvm/lib/Analysis/CMakeLists.txt:169]: GitHub is unhappy about the end of file here

**Review by arsenm — APPROVED** [review]

File: llvm/unittests/Analysis/CMakeLists.txt [hunk 13]
```diff
@@ -80,5 +80,11 @@ if(NOT WIN32)
   export_executable_symbols_for_plugins(AnalysisTests)
 endif()
 
+include(CheckCXXSymbolExists)
+check_cxx_symbol_exists(logf128 math.h HAS_LOGF128)
+if(HAS_LOGF128)
+  target_compile_definitions(AnalysisTests PRIVATE HAS_LOGF128)
```
Comment (nico) [inline llvm/unittests/Analysis/CMakeLists.txt:86]: Why is this needed?

It looks like nothing in AnalysisTests reads this
If it was needed in tests, shouldn't the tests include llvm-config.h and look at LLVM_HAS_LOGF128 instead anyways?

(Originally asked at https://github.com/llvm/llvm-project/commit/088aa81a545421933254f19cd3c8914a0373b493#r141559244 but here is probably better.)

**Comment by nickdesaulniers** [comment]
@efriedma-quic could we do something like this to solve https://github.com/llvm/llvm-project/issues/90685#issuecomment-2125416687 (if the target matched the host)?

**Comment by efriedma-quic** [comment]
See https://discourse.llvm.org/t/rfc-use-libms-logf128-for-constant-folding/81217/2

**Comment by davemgreen** [comment]
That sounds like the same issue we were trying to solve (depending on what logl means on the given target). There is a commit in https://github.com/llvm/llvm-project/pull/96287 and https://github.com/llvm/llvm-project/pull/104929 that got so very close, in that I believe it was working on all the buildbots. This can cause a really surprising perf swing on AArch64 where soft-float routines get called, so it would be good if we can get something fixed.
